core/src/main/java/org/elasticsearch/common/compress/CompressedXContent.java

```diff
@@ -37,23 +40,74 @@
  */
 public final class CompressedXContent {
 
+    private static int crc32(BytesReference data) {
+        OutputStream dummy = new OutputStream() {
+            @Override
+            public void write(int b) throws IOException {
+                // no-op
+            }
+            @Override
+            public void write(byte[] b, int off, int len) throws IOException {
+                // no-op
+            }
+        };
+        CRC32 crc32 = new CRC32();
+        try {
+            data.writeTo(new CheckedOutputStream(dummy, crc32));
+        } catch (IOException bogus) {
+            // cannot happen
+            throw new Error(bogus);
+        }
+        return (int) crc32.getValue();
+    }
+
     private final byte[] bytes;
-    private int hashCode;
+    private final int crc32;
 
-    public CompressedXContent(BytesReference data) throws IOException {
+    private CompressedXContent(BytesReference data, Integer crc32) throws IOException {
         Compressor compressor = CompressorFactory.compressor(data);
         if (compressor != null) {
             // already compressed...
             this.bytes = data.toBytes();
+            if (crc32 != null) {
+                this.crc32 = crc32;
+            } else {
+                this.crc32 = crc32(new BytesArray(uncompressed()));
+            }
         } else {
             BytesStreamOutput out = new BytesStreamOutput();
-            try (StreamOutput compressedOutput = CompressorFactory.defaultCompressor().streamOutput(out)) {
-                data.writeTo(compressedOutput);
+            try (OutputStream compressedOutput = CompressorFactory.defaultCompressor().streamOutput(out)) {
+                CRC32 checksum = null;
+                OutputStream checkedOutput = compressedOutput;
+                if (crc32 == null) {
+                    // compute the checksum if it was not provided
+                    checksum = new CRC32();
+                    checkedOutput = new CheckedOutputStream(compressedOutput, checksum);
+                }
+                data.writeTo(checkedOutput);
+                if (crc32 == null) {
+                    this.crc32 = (int) checksum.getValue();
+                } else {
+                    this.crc32 = crc32;
+                }
             }
             this.bytes = out.bytes().toBytes();
             assert CompressorFactory.compressor(new BytesArray(bytes)) != null;
         }
 
+        assert this.crc32 == crc32(new BytesArray(uncompressed()));
+    }
+
+    /**
+     * Create a new {@link CompressedXContent} with an already pre-computed
+     * crc32. This must be the crc32 of the <b>uncompressed</b> data.
+     */
+    public CompressedXContent(BytesReference data, int crc32) throws IOException {
```

Comment:
Instead of having this public ctor should we have one that has this signature:

```Java
public CompressedXContent(ToXContent xcontent, XContentType type, Params params) {
  // do the serialization here with checked output stream etc
}
```

that way we can hide the _CRC32_ impl detail entirely